hw/xbox/xid-sdl.c

```diff
@@ -511,18 +486,49 @@ static void usb_xbox_gamepad_realize(USBDevice *dev, Error **errp)
     SDL_JoystickEventState(SDL_ENABLE);
 #endif
 
-#ifdef FORCE_FEEDBACK
-    s->sdl_haptic = SDL_HapticOpenFromJoystick(s->sdl_gamepad);
-    if (s->sdl_haptic == NULL) {
-        fprintf(stderr, "Joystick doesn't support haptic\n");
-    } else {
-        if ((SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) == 0) {
-            fprintf(stderr, "Joystick doesn't support necessary haptic effect\n");
-            SDL_HapticClose(s->sdl_haptic);
-            s->sdl_haptic = NULL;
+    if (!SDL_HapticOpened(s->device_index)) {
+        s->sdl_haptic = SDL_HapticOpen(s->device_index);
+        if (s->sdl_haptic == NULL) {
+            fprintf(stderr, "Joystick doesn't support haptic effects\n");
+        } else {
+            if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) {
+                SDL_HapticEffect effect;
+                effect.leftright = (SDL_HapticLeftRight){
+                    .type = SDL_HAPTIC_LEFTRIGHT,
+                    .length = SDL_HAPTIC_INFINITY,
+                    .large_magnitude = 0x0000,
+                    .small_magnitude = 0x0000,
+                };
+
+                int effect_id = SDL_HapticNewEffect(s->sdl_haptic, &effect);
+                if (effect_id == -1) {
+                    const char* error = SDL_GetError();
+                    fprintf(stderr, "Failed to create haptic effect! SDL Error: %s", error);
+                    SDL_HapticClose(s->sdl_haptic);
+                    s->sdl_haptic = NULL;
+                    s->sdl_haptic_effect = NULL;
+                    return;
```

Comment:
return here is a crutch. It's convenient because this is the last branch in the method, but if more logic is added later this may cause tricky bugs. I will rework.

Reply:
Such things typically indicate you should be using functions